### PR TITLE
WIP: Function templates support

### PR DESCRIPTION
**Pull request by dbabokin**
This PR introduces function template support with syntax and semantic borrowed from C++. Specifically, keywords `template` and `typename` were added. They can be used to declare or define the function template:
```cpp
template <typename T> T foo(T t) { return t + 1; }
```

Template functions may be called through explicit type parameters specification syntax:
```cpp
void bar() {
    int i = foo<int>(1);
    uniform int ui = foo<uniform int>(1);
    float f = foo<float>(1.f);
}
```
While template argument deduction rules generally follow C++, there are some differences caused by existence of ``uniform``, ``varying`` and ``unbound`` types in ``ispc`` type system.  The template type parameter may resolve only to ``uniform`` and ``varying`` types, but not to ``unbound`` type. Consider the following example:
```cpp
template <typename T> T add(T a, T b) { return a + b; }

void foo() {
    // Note that these two lines call the same function:
    int i1 = add<int>(1, 2);                 // T = varying int
    varying int i2 = add<varying int>(1, 2); // T = varying int

    // And this call a uniform version:
    uniform int i3 = add<uniform int>(1, 2); // T = uniform int
}
```
The variability of template type parameter ``T`` may be overwritten by ``uniform`` and ``varying`` keywords, so ``uniform T`` and ``varying T``  are always valid types. But when ``uniform T`` and ``varying T`` are used to specify template function parameters it has an effect on template argument deduction process.  If variability keyword is specified and the type was successfully deduced, the default variability of type ``T`` is assumed to be the opposite of the variability keyword.  The logic behind it is that the keyword was specified "on purpose" to changed the variability of the type ``T``. Consider the following example:

```cpp
template <typename T> void foo1(T t);
template <typename T> void foo2(uniform T t);
template <typename T> void foo3(varying T t);

void bar() {
    uniform int ui;
    varying int vi;
    foo1(ui); // T is uniform int
    foo1(vi); // T is varying int
    foo2(ui); // T is varying int!
    foo2(vi); // error: varying type cannot be passed to uniform parameter
    foo3(ui); // T is uniform int!
    foo3(vi); // T is uniform int!
}
```
Template parameters type deduction is not yet implemented, but planned in this PR, i.e.:
```cpp
void bar() {
    int i = foo((varying int)1);
    uniform int ui = foo(1);
    float f = foo(programIndex + 1.f);
}
```

Also, this PR introduces the syntax for a set of function template features that are planned, but not yet implemented:
- default values for template type parameters
- non-type template type parameters (integer parameters)
- function specialization
The test demonstrating what is not supported is [here](https://github.com/dbabokin/ispc/blob/templates_new/tests/lit-tests/func_template_not_supported.ispc).

Here are the tests that demonstrate what is supported and working: [test1](https://github.com/dbabokin/ispc/blob/templates_new/tests/lit-tests/func_template_1.ispc), [test2](https://github.com/dbabokin/ispc/blob/templates_new/tests/lit-tests/func_template_2.ispc).

TODO for this PR:
- [x] type deduction support in the call expression to support `foo(arg1, arg2);` call syntax (partially completed)
- [x] support all types of expressions in template code (call expressions depended types are not supported yet)
- [x] user documentation update
- [x] design document
- [x] test coverage
- [ ] consider several refactorings and review memory allocations.

TODO for either this or follow up PRs:

- [ ] support for non-type template parameters - integer parameters.

**Feedback for this feature is very welcome.**
Linux and Windows builds of the latest version can be found in Appveyor artifacts. Current (as of writing this comment) Linux build is [here](https://ci.appveyor.com/api/buildjobs/j858ese3ok5ffrqa/artifacts/build%2Fispc-trunk-linux.tar.gz). Windows build is [here](https://ci.appveyor.com/api/buildjobs/s5qdooa3ps58aitt/artifacts/build%2Fispc-trunk-windows.zip).

### CODE REVIEW

File: src/func.h [hunk 1]
```diff
@@ -90,4 +90,12 @@ class TemplateParms {
     std::vector<const TemplateTypeParmType *> parms;
 };
 
+class TemplateArgs {
+  public:
+    TemplateArgs(const std::vector<std::pair<const Type *, SourcePos>> &args);
+    bool IsEqual(TemplateArgs &otherArgs) const;
+
+    std::vector<std::pair<const Type *, SourcePos>> args;
```
Comment (aneshlya) [inline src/func.h:100]: Should it be a private field?

File: src/func.h [hunk 2]
```diff
@@ -90,4 +90,12 @@ class TemplateParms {
     std::vector<const TemplateTypeParmType *> parms;
 };
 
+class TemplateArgs {
```
Comment (aneshlya) [inline src/func.h:95]: A comment on the class description would be great here.

File: src/type.h [hunk 3]
```diff
@@ -151,6 +151,9 @@ class Type {
         unbound. */
     bool HasUnboundVariability() const { return GetVariability() == Variability::Unbound; }
 
+    /** Resolves dependent type with mapping of template types to concrete type passed as argeument */
```
Comment (aneshlya) [inline src/type.h:157]: argeument-argument

File: src/ast.h [hunk 4]
```diff
@@ -134,6 +134,8 @@ class ASTNode {
         enumerant values defined in ispc.h. */
     virtual int EstimateCost() const = 0;
 
+    virtual ASTNode *Instantiate(TemplateInstantiation &templInst) const = 0;
```
Comment (aneshlya) [inline src/ast.h:137]: Comment with method description is needed here.

**Comment by mattgodbolt** [comment]
This branch is now built and installed every night on Compiler Explorer: https://godbolt.org/z/5959M8o8c (thanks to @pbrubaker for poking me)

File: src/type.h [hunk 5]
```diff
@@ -73,15 +73,16 @@ struct Variability {
...
+    ATOMIC_TYPE,              // 0
+    ENUM_TYPE,                // 1
+    POINTER_TYPE,             // 2
+    ARRAY_TYPE,               // 3
+    VECTOR_TYPE,              // 4
+    STRUCT_TYPE,              // 5
+    UNDEFINED_STRUCT_TYPE,    // 6
+    REFERENCE_TYPE,           // 7
+    FUNCTION_TYPE,            // 8
+    TEMPALATE_TYPE_PARM_TYPE, // 9
```
Comment (aneshlya) [inline src/type.h:85]: Typo TEMPALATE_TYPE_PARM_TYPE->TEMPLATE_TYPE_PARM_TYPE

File: src/stmt.cpp [hunk 6]
```diff
@@ -344,6 +356,18 @@ void DeclStmt::Print(Indent &indent) const {
...
+DeclStmt *DeclStmt::Instantiate(TemplateInstantiation &templInst) const {
+    std::vector<VariableDeclaration> instVars;
+    for (auto var : vars) {
+        Expr *instInit = var.init ? var.init->Instantiate(templInst) : nullptr;
+        Symbol *instSym = templInst.InstantiateSymbol(var.sym);
+        VariableDeclaration instDecl(instSym, instInit);
+        instVars.push_back(instDecl);
+    }
+    return new DeclStmt(instVars, pos);
+    ;
```
Comment (aneshlya) [inline src/stmt.cpp:368]: Extra semicolon

File: src/stmt.cpp [hunk 7]
```diff
@@ -344,6 +356,18 @@ void DeclStmt::Print(Indent &indent) const {
 
 int DeclStmt::EstimateCost() const { return 0; }
 
+DeclStmt *DeclStmt::Instantiate(TemplateInstantiation &templInst) const {
+    std::vector<VariableDeclaration> instVars;
+    for (auto var : vars) {
```
Comment (aneshlya) [inline src/stmt.cpp:361]: Is it more efficient to use `auto&`?

File: src/type.cpp [hunk 8]
```diff
@@ -627,6 +629,23 @@ const Type *TemplateTypeParmType::GetAsSOAType(int width) const {
...
 }
 
+const Type *TemplateTypeParmType::ResolveDependence(TemplateInstantiation &templInst) const {
+    const Type *resolvedType = templInst.InstantiateType(GetName());
+    if (resolvedType == NULL) {
+        Assert(m->errorCount > 0);
+        return NULL;
+    }
+    Variability v = variability;
+    // TODO: review this.
```
Comment (aneshlya) [inline src/type.cpp]: Still needs review?

File: src/func.cpp [hunk 9]
```diff
@@ -878,6 +878,24 @@ Symbol *TemplateInstantiation::InstantiateSymbol(Symbol *sym) {
     return instSym;
 }
 
+Symbol *TemplateInstantiation::InstantiateTemplateSymbol(TemplateSymbol *sym) {
+    // The function is assumed to be called once per instnatiation and
```
Comment (aneshlya) [inline src/func.cpp:1047]: instnatiation-> instantiation

File: src/expr.cpp [hunk 10]
```diff
@@ -3855,6 +3932,8 @@ Expr *FunctionCallExpr::TypeCheck() {
     if (func == NULL || args == NULL)
         return NULL;
 
+    // TODO: ???
```
Comment (aneshlya) [inline src/expr.cpp:3935]: ???

File: src/type.h (hunk 5 above)
Comment (dbabokin) [reply src/type.h:85]: Nice catch, fixed.

File: src/stmt.cpp (hunk 6 above)
Comment (dbabokin) [reply src/stmt.cpp:368]: Fixed.

File: src/stmt.cpp (hunk 7 above)
Comment (dbabokin) [reply src/stmt.cpp]: Fixed.

File: src/type.cpp (hunk 8 above)
Comment (dbabokin) [reply src/type.cpp]: It's gone by the final commit in the PR.

File: src/func.cpp (hunk 9 above)
Comment (dbabokin) [reply src/func.cpp:1049]: Fixed.

File: src/expr.cpp (hunk 10 above)
Comment (dbabokin) [reply src/expr.cpp:3935]: It's not a todo anymore, removed.

**Pull request merged**